// views/InteractiveOverlay2d.js
var THREE = require('three');
var manageSceneOnEnterFrameObjects = require('utils/manageSceneOnEnterFrameObjects');
var manageSceneOnResizeObjects = require('utils/manageSceneOnResizeObjects');
var manageSceneRayCollisions = require('utils/manageSceneRayCollisions');

function InteractiveOverlay2d(params) {
	var scene = new THREE.Scene();
	var camera = new THREE.OrthographicCamera(0, window.innerWidth, -window.innerHeight, 0, -100, 100);
	scene.add(camera);
	scene.camera = camera;

	manageSceneOnResizeObjects(scene);
	manageSceneOnEnterFrameObjects(scene);
	manageSceneRayCollisions(scene, params.rayEvents);

	this.camera = camera;
	this.scene = scene;
	this.onResize = this.onResize.bind(this);
}

function onResize(w, h) {
	if(w === undefined) return;
	this.scene.onResize(w, h);
	this.camera.right = w;
	this.camera.top = h;
	this.camera.updateProjectionMatrix();
}

function onEnterFrame(dt) {
	this.scene.onEnterFrame(dt);
}

function render(renderer) {
	renderer.render(this.scene, this.camera);
}

InteractiveOverlay2d.prototype = {
	onEnterFrame,
	onResize,
	render
}

module.exports = InteractiveOverlay2d;
