// cameras/ThirdPerson.js
const settings = require('settings');
const Blindfold = require('gameObjects/Blindfold');
const defaults = require('lodash.defaults');

var __defaultParams = {
  fov: 120,
  near: 0.01,
  far: 100,
  teleportDuration: 0.5
};

function ThirdPersonCamera (params) {
  params = defaults(params, __defaultParams);
  THREE.PerspectiveCamera.call(this, params.fov, 1, params.near, params.far);
  const blindfold = new Blindfold({
    radius: 0.2,
    color: 0x000000
  });
  this.add(blindfold);
  blindfold.animateOut();
  this.position.set(15, 15, 15);
  this.lookAt(new THREE.Vector3(0, 18, 0));
  var vrControls = new THREE.VRControls(this);
  vrControls.standing = true;
  this.vrControls = vrControls;
  this.blindfold = blindfold;
}

ThirdPersonCamera.prototype = Object.create(THREE.PerspectiveCamera.prototype);

var __proto = ThirdPersonCamera.prototype;
__proto.onEnterFrame = onEnterFrame;
__proto.onResize = onResize;

function onEnterFrame (dt) {
  // this.cameraControls.update();
}

function onResize (width, height) {
  const aspect = width / height;
  this.aspect = aspect;
  this.updateProjectionMatrix();
}

module.exports = ThirdPersonCamera;
