// cameras/ThirdPerson.js
const settings = require('settings');
const Blindfold = require('gameObjects/Blindfold');
const defaults = require('lodash.defaults');

var __defaultParams = {
  fov: 30,
  near: 0.01,
  far: 100,
  nearBG: 10,
  farBG: 2000,
  teleportDuration: 0.5
};

function ThirdPersonCamera (params) {
  params = defaults(params, __defaultParams);
  THREE.PerspectiveCamera.call(this, params.fov, 1, params.near, params.far);
  var cameraBG = new THREE.PerspectiveCamera(params.fov, 1, params.nearBG, params.farBG);
  cameraBG.layers.set(3);
  this.add(cameraBG);
  const blindfold = new Blindfold({
    radius: 0.2,
    color: 0x000000
  });
  this.add(blindfold);
  blindfold.animateOut();
  this.position.set(10, 10, 10);
  this.lookAt(new THREE.Vector3(-1, 1, 0));
  this.blindfold = blindfold;
  this.cameraBG = cameraBG;
}

ThirdPersonCamera.prototype = Object.create(THREE.PerspectiveCamera.prototype);

var __proto = ThirdPersonCamera.prototype;
__proto.onEnterFrame = onEnterFrame;
__proto.onResize = onResize;

function onEnterFrame (dt) {
  // this.cameraControls.update();
}

function onResize (width, height) {
  const aspect = width / height;
  this.aspect = aspect;
  this.updateProjectionMatrix();
}

module.exports = ThirdPersonCamera;
