// lights/Fire.js
var THREE = require('three');
var defaults = require('lodash.defaults');

var __defaultParams = {
	color: 0xdd7711,
	intensity: 1,
	distance: 4,
	decay: 1.2,
	wildIntensityRange: 0.5,
	wildIntensitySpeed: 0.1,
	wildMovementRange: 0.25,
	wildMovementSpeed: 0.3
}

function FireLight (params) {
	params = defaults(params, __defaultParams);
	THREE.Object3D.call(this);
	var light = new THREE.PointLight(params.color, params.intensity, params.distance, params.decay);
	this.add(light);
	this.wildIntensityRange = params.wildIntensityRange;
	this.wildIntensitySpeed = params.wildIntensitySpeed;
	this.wildMovementRange = params.wildMovementRange;
	this.wildMovementSpeed = params.wildMovementSpeed;
	this.light = light;
	this.intensity = params.intensity;
}

FireLight.prototype = Object.create(THREE.Object3D.prototype);

FireLight.prototype.onEnterFrame = function(dt) {
	var nextIntensity = this.intensity + (Math.random() - 0.5) * this.wildIntensityRange;
	this.light.intensity -= (this.light.intensity - nextIntensity) * this.wildIntensitySpeed;

	var nextX = (Math.random() - 0.5) * this.wildMovementRange;
	this.light.position.x -= (this.light.position.x - nextX) * this.wildMovementSpeed;
	var nextY = (Math.random() - 0.5) * this.wildMovementRange;
	this.light.position.y -= (this.light.position.y - nextY) * this.wildMovementSpeed;
	var nextZ = (Math.random() - 0.5) * this.wildMovementRange;
	this.light.position.z -= (this.light.position.z - nextZ) * this.wildMovementSpeed;
}

module.exports = FireLight;
