// gameObjects/GoldHornedSheep/index.js
var defaults = require('lodash.defaults');
var THREE = require('three');

require('three/examples/js/loaders/ColladaLoader2');

var __defaultParams = {
};
	
var __material;
function __getMaterial() {
	if(!__material) {
		__material = new THREE.MeshStandardMaterial({
			map: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-diffuse.png'),
			// normalMap: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-normal.png'),
			// emissive: 0xffffff,
			// emissiveMap: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-emissive.png'),
			subSurfaceAbsorbColorMap: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-subSurfaceAbsorbColor.png'),
			metalnessMap: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-metalness.png'),
			roughnessMap: THREE.ImageUtils.loadTexture('assets/textures/gold-horned-sheep-roughness.png'),
			// wireframe: true
		});
	}
	return __material;
}

function GoldHornedSheep(params) {
	params = defaults(params, __defaultParams);

	THREE.Object3D.call(this);

	this.material = __getMaterial();
	var loader = new THREE.ColladaLoader();
	loader.load('assets/meshes/gold-horned-sheep.dae', _onLoadColladaBuildMeshes.bind(this, params));
}

function _onLoadColladaBuildMeshes (params, colladaData) {
	var geometry = colladaData.library.geometries['Hedra001-lib'].build['gold horned sheep'].geometry;
	var sheep = new THREE.Mesh(
		geometry, 
		this.material
	);
	sheep.scale.multiplyScalar(0.03);
	sheep.rotation.x = Math.PI * -0.5;
	this.add(sheep);
	sheep.castShadow = true;
	sheep.receiveShadow = true;

}

GoldHornedSheep.prototype = Object.create(THREE.Object3D.prototype);

GoldHornedSheep.prototype.onEnterFrame = function(dt) {
	this.rotation.y += 0.01;
}

module.exports = GoldHornedSheep;
