// gameObjects/PlayerIntentionHighlightWalk.js
var THREE = require('three');
var defaults = require('lodash.defaults');

var __defaultParams = {
	color: 0xafbfef,
	height: 2
};

function PlayerIntentionHighlightWalk(params) {
	params = defaults(params, __defaultParams);
	var height = params.height;
	var geometry = new THREE.CylinderGeometry(0.5, 0.5, height, 16, 1, true);
	var colorFull = new THREE.Color(params.color);
	colorFull.multiplyScalar(0.5);
	var colorEmpty = new THREE.Color(0x000000);
	var heightHalf = height * 0.5;
	geometry.vertices.forEach(function(v){
		v.y += heightHalf;
	});
	geometry.faces.forEach(function(face) {
		[face.a, face.b, face.c].forEach(function(index) {
			var blend = geometry.vertices[index].y / height;
			face.vertexColors.push(blend > 0.5 ? colorEmpty : colorFull);
		});
	});
	var material = new THREE.MeshBasicMaterial({
		color: params.color,
		transparent: true,
		blending: THREE.AdditiveBlending,
		side: THREE.DoubleSide,
		depthWrite: false,
		vertexColors: THREE.VertexColors
	})
	THREE.Mesh.call(this, geometry, material);
}

PlayerIntentionHighlightWalk.prototype = Object.create(THREE.Mesh.prototype);

module.exports = PlayerIntentionHighlightWalk;
