// input/InteractiveDayDreamController.js
const geometryLibrary = require('geometry/library');
const tweener = require('tweeners/game-raf');
const ArmModel = require('vendor/ray-input-arm-model');
const urlParam = require('urlparam');

function InteractiveDayDreamController(params = {}) {
	THREE.Object3D.call(this);
	this.hasGamepadsAPI = 'getGamepads' in window.navigator;

	this.camera = params.camera;
	this.rayEvents = params.rayEvents;
	this.buttonPressed = false;

	this.onRaySelect = _onRaySelect.bind(this);
}

InteractiveDayDreamController.prototype = Object.create(THREE.Object3D.prototype);

var __q = new THREE.Quaternion();

InteractiveDayDreamController.prototype.onEnterFrame = function(dt) {
	if(!this.hasGamepadsAPI) {
		return;
	}
	// Loop over every gamepad and if we find any that have a pose use it.
	var vrGamepads = [];
	var gamepads = navigator.getGamepads();
	for (var i = 0; i < gamepads.length; ++i) {
		var gamepad = gamepads[i];
		// The array may contain undefined gamepads, so check for that as
		// well as a non-null pose.
		if (gamepad) {
			if (gamepad.pose) {
				vrGamepads.push(gamepad);
			}
			if ("hapticActuators" in gamepad && gamepad.hapticActuators.length > 0) {
				for (var j = 0; j < gamepad.buttons.length; ++j) {
					if (gamepad.buttons[j].pressed) {
						// Vibrate the gamepad using to the value of the button as
						// the vibration intensity.
						gamepad.hapticActuators[0].pulse(gamepad.buttons[j].value, 100);
						break;
					}
				}
			}
		}
	}
	var detectedController = vrGamepads.length > 0;
	_toggleRayListeners.call(this, detectedController);
	if(detectedController) {
		var gamepad = vrGamepads[0];
		__q.fromArray(gamepad.pose.orientation);
		if(!this.dayDreamController) {
			var dayDreamController = new THREE.Mesh(
				geometryLibrary.getCreator('dayDreamController')(),
				new THREE.MeshLambertMaterial({
					vertexColors: THREE.VertexColors,
					color: 0xffffff,
					opacity: 0,
					transparent: true,
				})
			);
			dayDreamController.visible = urlParam('showDayDreamController', true);
			tweener.to(
				dayDreamController.material,
				1,
				{
					opacity: 1
				}
			);

			this.add(dayDreamController);

			var armModel = new ArmModel();

			this.dayDreamController = dayDreamController;
			this.armModel = armModel;
		}
		var state = gamepad.buttons[0].pressed;
		if(this.buttonPressed !== state) {
			if(state) {
				this.rayEvents.dispatchOnSelect([0, 0]);
			}
			this.buttonPressed = state;
		}
	}
	if(this.dayDreamController) {
		this.armModel.setControllerOrientation(__q);
		this.armModel.setHeadOrientation(this.camera.quaternion);
		this.armModel.setHeadPosition(this.camera.position);
		this.armModel.update();
		this.quaternion.copy(this.armModel.pose.orientation);
		this.position.copy(this.armModel.pose.position);
	}
}

function _toggleRayListeners(state) {
	if(this.listeningForRays === state) return;
	this.listeningForRays = state;
	if(state) {
		this.rayEvents.addOnSelect(this.onRaySelect);
	} else {
		this.rayEvents.removeOnSelect(this.onRaySelect);
	}
}

function _onRaySelect(e) {
	debugger;
}
module.exports = InteractiveDayDreamController;
